### DISCUSSION TITLE
Wrong OpenAPI schema generation for optional query parameters

### DISCUSSION THREAD

**Discussion by tobias-feil-plana** [fastapi/fastapi#10912] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def read_root(
  test_param: Optional[str]
):
  return "Hello"
```

### Description
The code above generates the following query parameter definition in the openapi.json file:

```json
{
  "name": "test_param",
  "in": "query",
  "required": false,
  "schema": {
    "anyOf": [
      {
        "type": "string"
      },
      {
        "type": "null"
      }
    ],
    "title": "Test Param"
  }
},
```
This is not understood by Swagger, which in turn doesn't display the type in the UI:
![image](https://github.com/tiangolo/fastapi/assets/142786001/ff9deb97-56e9-4c22-a184-444e57ca5e78)

The following would suffice to make it work:

```json
{
  "name": "test_param",
  "in": "query",
  "required": false,
  "schema": {
    "type": "string",
    "title": "Test Param"
  }
}
```
As far as I understand, the `null` isn't necessary when `required` is `false` (which in turn isn't necessary either bc it's `false` by default but that's a question for a different day).

### Operating System

Linux

### Operating System Details

Ubuntu 22.04

### FastAPI Version

0.100.0

### Pydantic Version

2.5.2

### Python Version

Python 3.9.18

**Comment by JavierSanchezCastro** [comment]
Hi!

As explained here: https://fastapi.tiangolo.com/python-types/?h=optional#using-union-or-optional

Optional doesn't mean the parameter is optional, it means "could be None".

Apart from this, the picture you have included of the swagger does not seem to be the one generated by the code you have included (green background = POST endpoint) and the example is with GET.

Likewise, the code I think you are looking for is the following:

```python3
from fastapi import FastAPI
from pydantic.json_schema import SkipJsonSchema

app = FastAPI()

@app.get("/")
def read_root(test_param: str | SkipJsonSchema[None] = None):
  return "Hello"
```

Tested with
```
fastapi==0.108.0
pydantic==2.5.3
```

and

```
fastapi==0.100.0
pydantic==2.5.2
```

**Reply by tobias-feil-plana** [reply]
Thank you for the quick answer! :partying_face: 

You're right, it should have been declared as a post endpoint.

OK, I'll be happy to use `Union` over `Optional`, but the standard way of declaring an optional query param as defined in the docs [here](https://fastapi.tiangolo.com/tutorial/query-params/?h=query#optional-parameters) also doesn't render an openapi.json that shows the type in Swagger.

On the other hand, your way of doing it works and also looks good in Swagger. But 
1. this is not in the FastAPI docs; instead, they only mention the `Union[str, None]` way. But I would think that everyone would like their optional query parameters to be properly annotated in the Swagger UI
2. It seems cumbersome to write `SkipJsonSchema` when my use case is actually `Union[str, None]`. I would say the json schema should be skipped by default when defining an optional query param with `Union[str, None]`, or else their types will never show in the Swagger UI.

What do you think? Thanks again :slightly_smiling_face:

**Reply by JavierSanchezCastro** [reply, marked as answer]
Well, as I understand from here https://github.com/tiangolo/fastapi/discussions/10911#discussioncomment-8062800 and many other discussions about the same topic, it's something that needs to be fixed in the future, but something you can achieve using `SkipJsonSchema`

**Reply by tobias-feil-plana** [reply]
OK sorry for the hassle, I really searched for a while, hope that page will get better indexing on Google in the future :sweat_smile:
